Breaking change doc tool

ericstj opened the pull request:
This adds a workflow to run a script when a PR is labeled as `needs-breaking-change-doc-created`,

The script collects details about the PR, the breaking change issue template, and sample issues then asks an LLM to generate an issue based on that.

The issue content is then encoded into a clickable link which is then added as a comment to the PR.

I'm doing this from a dotnet branch so that I can hopefully create a PR off of it to demonstrate the workflow in action.

ericstj committed:
Initial check-in of breaking-change doc script

ericstj committed:
Adjust breaking change script for runtime repo

ericstj committed:
Remove label checking

ericstj committed:
Add a workflow triggered on breaking change docs needed.

ericstj committed:
Remove issue write permission

ericstj committed:
Simplify breaking change doc step

ericstj committed:
Use pull_request_target

ericstj committed:
Only run workflow on dotnet repo.

ericstj committed:
Use GitHub REST API for fetching from dotnet/docs

ericstj committed:
Fix log messages

ericstj committed:
Try models: read permission

rzikm commented on eng/breakingChanges/breaking-change-doc.ps1:
```diff
@@ -0,0 +1,985 @@
+    $encoded = $encoded -replace '"', '%22' -replace "'", '%27' -replace '<', '%3C' -replace '>', '%3E'
+
+    return $encoded
+}
```
Replace by `[System.Web.HttpUtility]::UrlEncode($text)`?

jeffhandley replied:
I think you might actually want [Uri.EscapeDataString](https://learn.microsoft.com/en-us/dotnet/api/system.uri.escapedatastring?view=net-9.0) here since you are only escaping a single element of data rather than the full Url itself.

rzikm commented on eng/breakingChanges/breaking-change-doc.ps1:
```diff
@@ -0,0 +1,985 @@
+                    $ghArgs += @("--system-prompt", $SystemPrompt)
+                }
+
+                $response = & gh @ghArgs
```
& is not required here


```suggestion
                $response = gh @ghArgs
```

rzikm commented on eng/breakingChanges/breaking-change-doc.ps1:
```diff
@@ -0,0 +1,985 @@
+                $requestJson = $body | ConvertTo-Json -Depth 10
+                $response = Invoke-RestMethod -Uri $endpoint -Method POST -Headers $headers -Body $requestJson
+
+                switch ($Config.LlmProvider) {
```
We already have a switch based on LlmProvider above, maybe we should separate REST calls implementations to separate methods per provider (and keep `Invoke-LlmApi` as a single point of entry)

rzikm reviewed (commented):
Looks great, I look forward to using this once the technical hurdles get ironed out.

ericstj committed:
Add support for Copilot CLI

ericstj committed:
Address feedback

ericstj commented:
I don't expect this to work, but also trying copilot CLI provider.

ericstj committed:
Fix yml indenting

ericstj committed:
Allow external API key for Github Models and Copilot CLI

Use models.
Make sure to disable tools when using copilot CLI

ericstj committed:
Use optional token when invoking gh models / copilot

ericstj committed:
Use GH_TOKEN instead of GITHUB_TOKEN

ericstj committed:
Merge branch 'main' into breakingChangeDocTool

jeffhandley commented on eng/breakingChanges/breaking-change-doc.ps1:
```diff
@@ -0,0 +1,1114 @@
+                # Get the merge commit for this PR
+                $mergeCommit = gh pr view $prNumber --repo $Config.SourceRepo --json mergeCommit --jq '.mergeCommit.oid' 2>$null
+
+                if ($mergeCommit) {
```
By limiting this to only run on merged PRs, we can drop a lot of this logic too and not worry about scenarios of walking arbitrary git histories.

jeffhandley commented on eng/breakingChanges/breaking-change-doc.ps1:
```diff
@@ -0,0 +1,1114 @@
+        }
+    }
+
+    # If no area labels found, fall back to file path analysis
```
What's the scenario for wanting to determine feature areas without area labels? I've found it's pretty time-consuming for the GH API to get the list of files from PRs, especially for PRs older that then newest 100 (presumably due to caching optimizations), so if we can avoid collecting the file names that would be beneficial.

jeffhandley replied:
Got it; thanks. I found the `dotnet-api-docs` repo as a good stress test for crawling PRs, and other repos where hundreds of files are edited at once on PRs commonly also exhibit the perf cliff I've seen. In order for [labeler-train.yml (dotnet/dotnet-api-docs)](https://github.com/dotnet/dotnet-api-docs/blob/main/.github/workflows/labeler-train.yml#L37) to work, it can only consume 1 pull request at a time and even with that, GitHub's API sometimes times out.

ericstj commented on .github/workflows/breaking-change-doc.yml:
```diff
@@ -0,0 +1,54 @@
+
+on:
+  pull_request_target:
+    types: [labeled]
```
True.  The reason I was thinking to run on label was so that we could toggle it to retrigger.  Maybe we can do that once closed too.

ericstj committed:
Address feedback

ericstj committed:
Only run the breaking change doc workflow on merged PRs

The workflow will now run when a PR is closed, or a label is appled.

In both cases it only proceeds if the PR has both been merged and has the breaking change doc label.

ericstj committed:
Add workflow dispatch to breaking change workflow

ericstj committed:
Add docs to breaking change workflow.

ericstj committed:
Fix workflow condition

ericstj commented:
it looks like the changes to URL escaping broke the formatting of newlines.  I'll need to see when that happened and undo it.

ericstj committed:
Fix calling gh CLI and use prompt file

ericstj committed:
Use powershell-yaml

ericstj committed:
Fix Clean to work correctly

ericstj committed:
Update readme

ericstj committed:
Fetch PR commits, and remove some redundant CLI calls.

jeffhandley commented on eng/breakingChanges/breaking-change-doc.ps1:
```diff
@@ -0,0 +1,1231 @@
+                "Comments: $($_.Comments.Count) comments available"
+            } else {
+                "No comments"
+            }
```
We recently hit upon a gap in our processes where we can have both issues and PRs labeled as https://github.com/dotnet/runtime/labels/breaking-change, but we missed auditing breaking change issues to ensure all closing PRs for them were also labeled as breaking.

I don't think there's anything to do _in this PR_ for that, but seeing the traversal of `ClosingIssues` here has me wondering if we should refine our approach for labeling--perhaps to automatically label merged PRs as breaking changes if a breaking-change issue was closed by the PR.

ericstj replied:
It would have to run whenever a label is applied to the issue - make it applied to the corresponding PR(s) whether open/closed.  If we can have the automation do that reliably it could fill that gap.

Another thing we could do would be to nag people when applying the label to an issue and tell them to apply to PR instead.

ericstj commented on .github/workflows/breaking-change-doc.yml:
```diff
@@ -0,0 +1,73 @@
+  generate-breaking-change-doc:
+    if: |
+      github.repository_owner == 'dotnet' && (
+        (github.event_name == 'pull_request_target' && github.event.action == 'closed' && github.event.pull_request.merged == true && contains(github.event.pull_request.labels.*.name, 'needs-breaking-change-doc-created')) ||
```
Just stole the pattern from our other workflows https://github.com/dotnet/runtime/blob/eeaef45074f577f35bfd1b6b02a029d4053a9683/.github/workflows/check-service-labels.yml#L20

ericstj commented on eng/breakingChanges/breaking-change-doc.ps1:
```diff
@@ -0,0 +1,1231 @@
+
+EXAMPLES:
+    .\breaking-change-doc.ps1 -PrNumber 114929                                                        # Process specific PR
+    .\breaking-change-doc.ps1 -Query "repo:dotnet/runtime state:closed label:needs-breaking-change-doc-created is:merged merged:>2024-09-16 -milestone:11.0.0"
```
That's a good point.  I want to get rid of the repo default eventually, but will probably make the repo be passed in separately.  We do pass this to the GH CLI already, so it's likely redundant to place in the query.

ericstj committed:
Remove repo from queries
